### ISSUE TITLE
failed to refresh the status of the execution by Retention job

### ISSUE THREAD

**Issue by drehpehs** [goharbor/harbor#17646]
We have a project with daily retention scheduled. 
This project has a lots of images, so it takes more time to consume the retention job for sure.

But today we found it is failed and the DB connection are occupied and reached the limitation.
The Harbor became unavailable because all the connection are full and can't be released.
Besides, we found the log of its retention is weird and different than other projects also with daily retention schedule.
It has two logs per day. 

And we got a lots of ERRORs in the log
/var/log/harbor/core.log
```
Oct 13 00:01:44 240.0.1.1 core[20264]: 2022-10-13T00:01:44Z [ERROR] [/lib/http/error.go:54]: {"errors":[{"code":"UNKNOWN","message":"unknown: failed to refresh the status of the execution 984786 after 5 retries"}]}
Oct 13 00:01:44 240.0.1.1 core[20264]: 2022-10-13T00:01:44Z [ERROR] [/lib/http/error.go:54]: {"errors":[{"code":"UNKNOWN","message":"unknown: failed to refresh the status of the execution 984786 after 5 retries"}]}
Oct 13 00:02:24 240.0.1.1 core[20264]: 2022-10-13T00:02:24Z [ERROR] [/lib/http/error.go:54]: {"errors":[{"code":"UNKNOWN","message":"unknown: failed to refresh the status of the execution 984786 after 5 retries"}]}
...
```
The execution id 984786 is about the project we mentioned.

/var/log/harbor/jobservice.log
```
Oct 13 00:01:44 240.0.1.1 jobservice[20264]: 2022-10-13T00:01:44Z [ERROR] [/jobservice/runtime/bootstrap.go:149]: trigger hook event error: {"errors":[{"code":"UNKNOWN","message":"internal server error"}]}
Oct 13 00:01:44 240.0.1.1 jobservice[20264]: 2022-10-13T00:01:44Z [ERROR] [/jobservice/runtime/bootstrap.go:149]: trigger hook event error: {"errors":[{"code":"UNKNOWN","message":"internal server error"}]}
Oct 13 00:02:24 240.0.1.1 jobservice[20264]: 2022-10-13T00:02:24Z [ERROR] [/jobservice/runtime/bootstrap.go:149]: trigger hook event error: {"errors":[{"code":"UNKNOWN","message":"internal server error"}]}
...
Oct 13 00:04:39 240.0.1.1 jobservice[20264]: 2022-10-13T00:04:39Z [ERROR] [/jobservice/hook/hook_agent.go:155]: Retry: sending hook event error: {"errors":[{"code":"UNKNOWN","message":"internal server error"}]}
...
```

We have 7158 repositories and use 417 GB in this project.
Is it because of the big project and reach the performance bottleneck?
Do you have any idea to improve/resolve this?
Thank you in advance.

Versions:
Please specify the versions of following systems.

harbor version: [2.2.2]
docker engine version: [18.09.0]
docker-compose version: [1.24.1]

https://github.com/goharbor/harbor/issues/14075
Not pretty sure is this related to this issue?

**Comment by drehpehs** [comment]
Hi, 

We found the number of repositories can affect the retention performance.
Even we have retention rule to clean up some redundant repositories, but it only delete the artifact. 
The repository with empty artifacts is still there and every time the retention job still check them one by one.

After we remove empty repositories, we decreased the repository number from thousands to hundreds. The retention time also downgrade the time from mins to secs in our test.
